iso15118/shared/states.py

```diff
@@ -290,26 +312,33 @@ def create_next_message(
             except ValidationError as exc:
                 logger.exception(exc)
                 raise exc
+            self.message = to_be_exi_encoded
+        elif isinstance(next_msg, Base64):
+            # Incoming message is base64 encoded EXI message.
+            # So a base64 decode should retrieve the EXI stream.
+            self.message = next_msg
+            exi_payload = base64.b64decode(next_msg.message)
         else:
             to_be_exi_encoded = next_msg
+            self.message = to_be_exi_encoded
 
-        self.next_msg = to_be_exi_encoded
-
-        # If either next_msg or next_msg_payload_type are None, the state's
-        # attribute next_v2gtp_msg will not be set. This causes the state
-        # machine to raise a FaultyStateImplementationError if next state is
-        # not set to Terminate, so no need to raise anything here.
-        if next_msg and next_msg_payload_type:
+        # If to_be_exi_encoded is None it is possible that exi_payload is already
+        # set (for eg:CertificateInstallationRes from backend).
+        # Otherwise, EXI encode the message.
+        if to_be_exi_encoded and next_msg_payload_type:
             # Step 3
-            exi_payload: bytes = bytes(0)
             try:
                 exi_payload = EXI().to_exi(to_be_exi_encoded, namespace)
             except EXIEncodingError as exc:
                 logger.error(f"{exc}")
                 self.next_state = Terminate
                 raise
 
-            # Step 4
+        # If exi_payload is None, next_v2gtp_msg will not be set.
+        # This in turn, will raise FaultyStateImplementationError if next state is
+        # not set to Terminate, so no need to raise anything here.
+        # Step 4
+        if exi_payload:
```

Comment:
if there is a `to_be_exi_encoded` then there will be a `exi_payload`

The only case where there is no `to_be_exi_encoded` is when we have a base64 message. but in that case we have a `exi_payload`.
This covers all the cases, so I dont think we need this check

Reply:
Hmm...actually in the case of base64 response `to_be_exi_encoded` is not set (as it is not to be exi_encoded) - in this case exi_payload is set directly by base64 decoding the response. This is why the additional `if exi_payload:` was added...

Reply:
correct, `to_be_exi_encoded` is not set, but `exi_payload` will be: https://github.com/SwitchEV/iso15118/pull/80/files#diff-021c1e2405f51ff9a35df9843901c0ffbfe3ea77a23a2bf2844826f7cf695335R320

so, the code in here: https://github.com/SwitchEV/iso15118/pull/80/files#diff-021c1e2405f51ff9a35df9843901c0ffbfe3ea77a23a2bf2844826f7cf695335R346
can run, isnt that right?

Reply:
Ah kk..correct... :) ...have removed the check.